pytorch_lightning/plugins/pipe_rpc_plugin.py

```diff
@@ -0,0 +1,302 @@
+    from fairscale.nn import LazyModule, PipeRPCWrapper
+    from fairscale.nn.model_parallel.utils import ensure_divisibility
+    from fairscale.nn.pipe import balance as pipe_balance
+    from fairscale.nn.pipe import rpc as rpc_pipe
+    from fairscale.nn.pipe.pipeline import PipelineStyle
+    from torch.distributed import rpc
+
+
+def get_worker_map():
+    # TODO, is this correct with multinodes?
+    return {rank: f"worker{rank}" for rank in range(torch_distrib.get_world_size())}
+
+
+def register_optimizers(ctx, model):
+    optimizers, lr_schedulers, optimizer_frequencies = model.trainer.init_optimizers(model)
+    model.trainer.optimizers = optimizers
+    model.trainer.lr_schedulers = lr_schedulers
+    model.trainer.optimizer_frequencies = optimizer_frequencies
+    model.trainer.convert_to_lightning_optimizers()
+    model.trainer.is_master = False
+
+
+def do_nothing_optimizer_closure():
+    return
+
+
+def run_optimizer(ctx, model):
+    trainer = model.trainer
+    opt_idx = ctx["opt_idx"]
+    optimizer = trainer.optimizers[opt_idx]
+    closure = getattr(optimizer, "_closure", do_nothing_optimizer_closure)
+    optimizer.step(closure=closure)
+
+
+def save(ctx, model):
+    num_gpus_per_model = ctx["num_gpus_per_model"]
+    rank = torch_distrib.get_rank()
+    if rank in range(num_gpus_per_model):
+        seq = list(model.children())[0]
+        torch.save(seq, f"seq_{rank}.pt")
+
+
+def reload_sequential(num_gpus_per_model):
+    partial_seqs = [torch.load(f"seq_{rank}.pt", map_location='cpu') for rank in range(num_gpus_per_model)]
+    seq = nn.Sequential()
+    for p_seq in partial_seqs:
+        for name, child in p_seq.named_children():
+            seq.add_module(name, child)
+    # delete tmp files
+    [os.remove(f"seq_{rank}.pt") for rank in range(num_gpus_per_model)]
+    return seq
+
+
+class LightningPipeModule(nn.Module):
+    """
+        This class wraps Fairscale Pipe and PipeRCPWrapper class.
+
+        Args:
+            module: nn.Sequential
+                sequential model to be balanced among several gpus
+
+            balance: list of ints
+                list of number of layers in each partition.
+
+            checkpoint (str) = 'never'
+                when to enable checkpointing, one of ``'always'``,
+                ``'except_last'``, or ``'never'`` (default: ``'except_last'``)
+
+            balance_mode: str = "balance_by_size"
+                when balance is not provided, the model can be balanced either by size or time.
+                refer to balance description.
+
+            mode: PipeMode
+                the mode enables switching between Pipe and PipeRCPWrapper class
+    """
+    def __init__(self,
+                 module: nn.Sequential,
+                 balance: List[int],
+                 microbatches: int = 8,
+                 checkpoint='never',
+                 pipe_cls=None):
+        super().__init__()
+        self.module = module
+        self.balance = balance
```

Comment:
Can you add a check here whether the total number of layers in `balance` matches the number of layers in `module`? What happens if it doesn't?

Reply:
Great catch !